### ISSUE TITLE
The Jacobian for Finite Strain in tensor_mechanics is incorrect

### ISSUE THREAD

**Issue by tonkmr** [idaholab/moose#7028]
### Description of the enhancement or error report
In the tensor_mechanics module, the Jacobian for all finite strain problems is only an approximation. Even if the Jacobian_mult passed in by the ComputeStress material is right, the Jacobian is still an approximation because the derivative of the finite strain with respect to the gradient of the displacements is not considered. So, finite strain problems should NEVER be run with solve = newton right now.

### Rationale for the enhancement or information for reproducing the error
If we can improve the Jacobian or even make it exact, the solves will be greatly improved

### Identified impact
I think we might need to consider a change to how the Jacobian is calculated in all the StressDivergence kernels. Currently, we pass a fourth order tensor to the kernels, which is used to determine the Jacobian. However, it might make more sense to pass in a second order tensor. If u_i is the nodal values of the displacement vector:

dstress/du_i = dstress/dstrain*dstrain/dgrad_u*dgrad_u/du_i.

So, we could have dstress/dstrain be a rank 2 tensor calculated by the ComputeStress material and dstrain/dgrad_u be a rank 2 tensor calculated by the ComputeStrain material.

Does this make sense?

**Comment by tonkmr** [comment]
@dschwen @cpritam @jasondhales @bwspenc @sapitts @amjokisaari I would love to hear your thoughts on this.

**Comment by amjokisaari** [comment]
If I'm following correctly, you're essentially doing a chain rule to incorporate the gradients of the displacements, right?

Generally speaking I think it's a good idea to make code work as well as possible with the various combinations of MOOSE solve types/solid mechanics models/etc.  You don't want users to be caught in "gotcha" situations. 

Making this switch would require some pretty solid documentation, both on the mailing list to tell people of the changes, and on the wiki page to make sure people know how they need to update their code (if they've made their own materials for stresses and strains, like yours truly ;) )

**Comment by cpritam** [comment]
@tonkmr If I am right you are mentioning about models that use deformation gradient (F). For such models we can obtain dstress/dF (4th order) from computestress. The trick part will be to get dF=d(grad_u) (grad_u in current config). More straight forward will be dF=d(Grad_u). But we can break symmetries, which anyways is not assumed in the FE assembly. The second part can be obtained from computestrain.

**Comment by tonkmr** [comment]
@cpritam you are correct, the dstress/dstrain piece should be 4th order, which is our Jacobian_mult that we use now. I have fixed that above. 

If we take the simplest finite strain case, ComputeFiniteStrainElasticStress, the 

stress = C x strain_increment, where the strain increment is calculated using

strain_increment = -Cinv_I * 0.5 + Cinv_I * Cinv_I * 0.25;

So, here we have Jacobian_mult = C, but we don't currently include dstrain_increment/d_grad_u part. I'm just wondering if it is worth trying to get some estimate of this piece.

**Comment by cpritam** [comment]
I think it would improve the behavior. I more thing we may consider is the
effect of F_bar. Right now thats neglected as well.

Pritam

**Comment by WilkAndy** [comment]
Heheheh, i think we all must be getting old because we (or most of us!) discussed this sometime back when doing the consistent tangent operator in plasticity.  To get the correct jacobian we needed that, but also the finite-strain stuff.  I'd be very appreciative if someone did this, as for me it will DEFINITELY make a difference.

**Comment by tonkmr** [comment]
I am definitely getting old. Plus, I have been running some plasticity simulations and the pore convergence is getting on my nerves. Its much better when I use a better Jacobian from the FDP preconditioner,

**Comment by WilkAndy** [comment]
Yea, basically i only run in small-strain mode because of the jacobian.  So @tonkmr , are you volunteering to code this up?

**Comment by tonkmr** [comment]
@cpritam we should work together to try to improve the Jacobian.

**Comment by cpritam** [comment]
@tonkmr That sounds good to me. If you want we can discuss offline.

**Comment by WilkAndy** [comment]
Thanks guys - it'll be very valuable to me.
